Documentation for COVID-19 services (#2423)
alphagov/govuk-developer-docs: source/manual/covid-19-services.html.md

- [Service](https://coronavirus-vulnerable-people.service.gov.uk/live-in-england)
- [GitHub](https://github.com/alphagov/govuk-coronavirus-vulnerable-people-form)

## Architecture

Both applications have a similar architecture of:

- **Framework:** Ruby on Rails
- **Database:** AWS RDS (business) and DynamoDB (vulnerable people)
- **Hosting:** PaaS
- **CDN:** AWS CloudFront
- **CI:** Travis
- **CD:** Concourse
- **Logging:** Sentry and Logit

### Application structure

Each application is a standard Rails application with:

- question pages, each has their own controller, view and route
- a check your answers page
- a confirmation ("Thank you") page
- ineligible pages (vulnerable people form only)
- a privacy page

To run one of the applications locally, see the README in the GitHub repo.

## Session and data storage

### Session storage

While the user is filling out the form, we use session storage to
store the user's data.

For the vulnerable people form, we store session data in an encrypted
cookie, persisted in the browser and sent back to the server for every
question.  Cookies have a limit of 4KB, so this approach could cause
errors if the user submits large inputs.

> For the business volunteering form, we store the following for 2

Reviewer: Very minor: The vulnerable people form only stores session data for 2 hours as well, but I'm not sure if that needs to be recorded here.

Reviewer 2: ```suggestion
For both the business volunteering and vulnerable people forms, we store the following for 2
```

changes to application code on the master branch are continuously
deployed to staging.  Smoke tests are set up to run on staging, if
these pass the applications are automatically deployed to production.
The smoke tests are the feature tests within the application
repository.

The Concourse deployment pipeline and task configuration for each
application is stored in the "concourse" directory at the root of each
repository.

Links to Concourse Pipelines:

- [`govuk-corona-business-volunteer-form`](https://cd.gds-reliability.engineering/teams/govuk-tools/pipelines/govuk-corona-business-volunteer-form)
- [`govuk-corona-vulnerable-people-form`](https://cd.gds-reliability.engineering/teams/govuk-tools/pipelines/govuk-corona-vulnerable-people-form)

For access login with your GitHub account, you need to be a part of
the GOV.UK Production team.  You'll also need to be on the VPN.

### Administering Pipelines

To administer pipelines you must have the `pipeline-operator` role.
Pipelines are administered using the `fly` CLI tool.  `fly` can be
downloaded directly from the Concourse dashboard using the links in
the bottom bottom right of the screen (you probably want to click the

> apple).

Reviewer: For me, `fly` downloads as a text file.

Reviewer 3: You can `brew cask install fly`, too.  Thanks to @benthorner for finding that one!

```suggestion
apple). If that doesn't download an executable file, you can install with `brew cask install fly`.
```

This service allows businesses to tell us how they might be able to
help with the response to coronavirus.  This may include goods or
services such as medical equipment, hotel rooms or childcare.

- [Start page](https://www.gov.uk/coronavirus-support-from-business)
- [Service](https://coronavirus-business-volunteers.service.gov.uk/medical-equipment)
- [GitHub](https://github.com/alphagov/govuk-coronavirus-business-volunteer-form)

## Get coronavirus support as an extremely vulnerable person

This service allows citizens identified as vulnerable by the NHS to
tell us if they need help accessing essential supplies and support.
Users will have received a link to the service in a letter or a text
message from the NHS, or been advised by their GP to fill in the form.
They can fill the form in themselves, or someone else can fill it in
for them.

- [Start page](https://www.gov.uk/coronavirus-extremely-vulnerable)
- [Service](https://coronavirus-vulnerable-people.service.gov.uk/live-in-england)
- [GitHub](https://github.com/alphagov/govuk-coronavirus-vulnerable-people-form)

## Architecture

Both applications have a similar architecture of:

- **Framework:** Ruby on Rails

> - **Database:** AWS RDS

Reviewer 4: This is still DynamoDB for the vulnerable people form

technical documentation][paas-docs].

#### Get access

To get access, ask an organisation manager to invite you to the
`govuk_development` organisation.  They need to give you a role that
lets you access both `staging` and `production` spaces and the
applications inside.

#### Log in

Before you start, [install the CloudFoundry CLI][cf-cli].

1. Log in with `cf login -a api.cloud.service.gov.uk --sso`
2. Select staging or production with `cf target -s <staging or production>`

Both applications are Rails applications, and use the
[`ruby_buildpack`][] to build and deploy to PaaS.  You can read more
about [managing PaaS applications][paas-managing] and [monitoring PaaS
applications][paas-monitoring].

#### Manage backing services

The business volunteer application uses PaaS's [Redis backing
service][paas-redis].  The backing service is attached to the
application by a setting in the `manifest.yml` file in the repository
root.

The JSON data in the `VCAP_SERVICES` environment variable exposes
settings and credentials for the backing service automatically.

### AWS

We have two AWS accounts (staging and production) for CDN and data

> storage. They contain RDS / DynamoDB, IAM and CloudFront resources,

Reviewer 4: ```suggestion
storage. They contain DynamoDB, IAM and CloudFront resources,
```
The RDS instance is managed via PaaS

application is stored in the "concourse" directory at the root of each
repository.

Links to Concourse Pipelines:

- [`govuk-corona-business-volunteer-form`](https://cd.gds-reliability.engineering/teams/govuk-tools/pipelines/govuk-corona-business-volunteer-form)
- [`govuk-corona-vulnerable-people-form`](https://cd.gds-reliability.engineering/teams/govuk-tools/pipelines/govuk-corona-vulnerable-people-form)

For access login with your GitHub account, you need to be a part of
the GOV.UK Production team.  You'll also need to be on the VPN.

### Administering Pipelines

To administer pipelines you must have the `pipeline-operator` role.
Pipelines are administered using the `fly` CLI tool.  `fly` can be
downloaded directly from the Concourse dashboard using the links in
the bottom bottom right of the screen (you probably want to click the
apple).

To get started, firstly configure a target and log in:

```
fly --target cd-govuk-tools login --team-name govuk-tools --concourse-url https://cd.gds-reliability.engineering
```

Changes to the pipeline have to be applied manually using the
`set-pipeline` command.  To apply changes, run the following command

> from the project root:

Reviewer 4: Maybe add a line "Committing the changes to master will not update the pipeline." to be clear?

`govuk_development` organisation.  They need to give you a role that
lets you access both `staging` and `production` spaces and the
applications inside.

#### Log in

Before you start, [install the CloudFoundry CLI][cf-cli].

1. Log in with `cf login -a api.cloud.service.gov.uk --sso`
2. Select staging or production with `cf target -s <staging or production>`

Both applications are Rails applications, and use the
[`ruby_buildpack`][] to build and deploy to PaaS.  You can read more
about [managing PaaS applications][paas-managing] and [monitoring PaaS
applications][paas-monitoring].

#### Manage backing services

The business volunteer application uses PaaS's [Redis backing
service][paas-redis].  The backing service is attached to the
application by a setting in the `manifest.yml` file in the repository
root.

The JSON data in the `VCAP_SERVICES` environment variable exposes
settings and credentials for the backing service automatically.

### AWS

We have two AWS accounts (staging and production) for CDN and data
storage. They contain DynamoDB, IAM and CloudFront resources,
which were provisioned by Terraform in the [tech-ops-private
repository](techops-repo).

To log in to AWS:

> 1. Install `gds-cli` v2.9.0 or above:

Reviewer 3: Ooh, ooh, if we merge #2415 soon we can link to that for "how to install the GDS CLI".

Reviewer 2: Resolved here: https://github.com/alphagov/govuk-developer-docs/pull/2423#discussion_r398664510

This service allows businesses to tell us how they might be able to
help with the response to coronavirus.  This may include goods or
services such as medical equipment, hotel rooms or childcare.

- [Start page](https://www.gov.uk/coronavirus-support-from-business)
- [Service](https://coronavirus-business-volunteers.service.gov.uk/medical-equipment)
- [GitHub](https://github.com/alphagov/govuk-coronavirus-business-volunteer-form)

## Get coronavirus support as an extremely vulnerable person

This service allows citizens identified as vulnerable by the NHS to
tell us if they need help accessing essential supplies and support.
Users will have received a link to the service in a letter or a text
message from the NHS, or been advised by their GP to fill in the form.
They can fill the form in themselves, or someone else can fill it in
for them.

- [Start page](https://www.gov.uk/coronavirus-extremely-vulnerable)
- [Service](https://coronavirus-vulnerable-people.service.gov.uk/live-in-england)
- [GitHub](https://github.com/alphagov/govuk-coronavirus-vulnerable-people-form)

## Architecture

Both applications have a similar architecture of:

- **Framework:** Ruby on Rails

> - **Database:** AWS RDS (business) and DynamoDB (vulnerable people)

Reviewer 5: ```suggestion
- **Database:** AWS RDS (business volunteering) and DynamoDB (vulnerable people)
```

- **Hosting:** PaaS
- **CDN:** AWS CloudFront
- **CI:** Travis
- **CD:** Concourse
- **Logging:** Sentry and Logit

### Application structure

Each application is a standard Rails application with:

- question pages, each has their own controller, view and route
- a check your answers page
- a confirmation ("Thank you") page
- ineligible pages (vulnerable people form only)
- a privacy page

To run one of the applications locally, see the README in the GitHub repo.

## Session and data storage

### Session storage

While the user is filling out the form, we use session storage to
store the user's data.

> For the vulnerable people form, we store session data in an encrypted

Reviewer: ```suggestion
For the vulnerable people form, we store session data for 2 hours in an encrypted
```

Reviewer 2: done!
will need to amend this to 4 hours when this card is complete: https://trello.com/c/Zuq2NJO6/198-increase-session-expiry-from-2-to-4-hours

cookie, persisted in the browser and sent back to the server for every
question.  Cookies have a limit of 4KB, so this approach could cause
errors if the user submits large inputs.

> For both the business volunteering and vulnerable people forms, we store the following for 2

Reviewer: ```suggestion
For the business volunteering form, we store the following for 2
```

Reviewer 2: amended: https://github.com/alphagov/govuk-developer-docs/pull/2423/commits/808d17ce74025f746d4c24b88c7c2589a7b69322

hours:

- the session id in an encrypted cookie

> - the session data in Redis

Reviewer: As mentioned above, for now session data for the vulnerable people form is stored in a cookie, not Redis, but this may change in the next few days.

Reviewer 3: Would you like to rewrite this section, Leena? 😇

Reviewer 2: amended this here: https://github.com/alphagov/govuk-developer-docs/pull/2423/commits/808d17ce74025f746d4c24b88c7c2589a7b69322

### Data storage

For both applications, when the user submits the form on the "Check
your answers" page, the application writes user data to the database.

> The application only has permission to write items to the database.

Reviewer 5: This is only the case for the vulnerable people form.  The business volunteering form has read access to the Postgres database too.

Reviewer 2: I've got a PR coming up for that.

Reviewer 2: [New commit here](https://github.com/alphagov/govuk-developer-docs/pull/2423/commits/465f1c6a1ac4306940069b076127083db843e7fd)

For security and privacy reasons, there's no way to read or change

> already-submitted user data.

Reviewer 5: Only for the vulnerable people form.

Reviewer 2: I've got a PR coming up for that.

Reviewer 2: [New commit here](https://github.com/alphagov/govuk-developer-docs/pull/2423/commits/465f1c6a1ac4306940069b076127083db843e7fd)

> Developers should not have access to the production database.

Reviewer 5: Same as my previous comments, only for the vulnerable people form.  The business volunteering database is being treated like any other GOV.UK personal data store (e.g. signon or email-alert-api).

Reviewer 2: [New commit here](https://github.com/alphagov/govuk-developer-docs/pull/2423/commits/465f1c6a1ac4306940069b076127083db843e7fd)

- [`govuk-coronavirus-vulnerable-people-form`](https://govuk-coronavirus-vulnerable-people-form-stg.cloudapps.digital/live-in-england)

These are protected by HTTP basic auth.  You can find the credentials
in govuk-secrets and get them by running:

```
cd ~/govuk/govuk-secrets/pass
./edit.sh 2ndline coronavirus-forms/creds
```

## Monitoring

There's a [Splunk dashboard][splunk] for both of these services.  To
access Splunk, you need to have the `GDS-006-GOVUK` permission on your
Google account.  To get this permission, raise an IT Helpdesk ticket
and post in `#cyber-security-help` to get them to confirm the request
is legitimate.

Application errors are sent to the GOV.UK hosted Sentry:

- [`govuk-coronavirus-business-volunteer-form`](https://sentry.io/organizations/govuk/issues/?project=5172100)
- [`govuk-coronavirus-vulnerable-people-form`](https://sentry.io/organizations/govuk/issues/?project=5170680)

These are under the GOV.UK Sentry account and you can sign in with
your Google Account.

Logs for both applications are streamed to Logit and can be found in
following dashboards:

GOV.UK Production Corona Forms *TODO: where?*

> GOV.UK Staging Corona Forms *TODO: where?*

Reviewer 3: ```suggestion
GOV.UK Production Corona Forms](https://logit.io/a/1c6b2316-16e2-4ca5-a3df-ff18631b0e74/s/04b46992-f653-4c14-965c-236e9a6c2777/kibana/access)
[GOV.UK Staging Corona Forms](https://logit.io/a/1c6b2316-16e2-4ca5-a3df-ff18631b0e74/s/7a0be476-6535-4544-8318-4c7a130948e8/kibana/access)
```

> GOV.UK Production Corona Forms](https://logit.io/a/1c6b2316-16e2-4ca5-a3df-ff18631b0e74/s/04b46992-f653-4c14-965c-236e9a6c2777/kibana/access)

Reviewer: This link is missing the opening `[`

Reviewer 3: ```suggestion
[GOV.UK Production Corona Forms](https://logit.io/a/1c6b2316-16e2-4ca5-a3df-ff18631b0e74/s/04b46992-f653-4c14-965c-236e9a6c2777/kibana/access)
```

---
owner_slack: "#govuk-corona-forms-tech"
title: COVID-19 Services
section: Services
layout: manual_layout
parent: "/manual.html"
last_reviewed_on: 2020-03-25
review_in: 1 month
---

We recently launched 2 new services:

- Offer coronavirus support from your business
- Get coronavirus support as an extremely vulnerable person

> There is also an interactive voice response (IVR) automated phone

Reviewer 2: It appears TechOps maintains this, ongoing chat in slack about if they have any docs we can link to.

## Session and data storage

### Session storage

While the user is filling out the form, we use session storage to
store the user's data.

For the vulnerable people form, we store session data in an encrypted
cookie, persisted in the browser and sent back to the server for every
question.  Cookies have a limit of 4KB, so this approach could cause
errors if the user submits large inputs.

For the business volunteering form, we store the following for 2
hours:

- the session id in an encrypted cookie
- the session data in Redis

### Data storage

For both applications, when the user submits the form on the "Check
your answers" page, the application writes user data to the database.
The application only has permission to write items to the database.
For security and privacy reasons, there's no way to read or change
already-submitted user data.

Developers should not have access to the production database.

## Hosting

### Paas

Both applications are hosted on GOV.UK Platform as a Service (PaaS) in
the Ireland region.  PaaS provides a scalable hosting platform based
on CloudFoundry managed by GDS.  You can read more in the [PaaS
technical documentation][paas-docs].

#### Get access

> To get access, ask an organisation manager to invite you to the

Reviewer 6: This line could specify how/provide a link to find out who an organisation manager is.

Reviewer 7: The org managers are the GOV.UK Senior Tech team, so what about our group email address? govuk-senior-tech-members@digital.cabinet-office.gov.uk

Reviewer 3: ```suggestion
To get access, email govuk-senior-tech-members@digital.cabinet-office.gov.uk and ask for an invite to the
```

Concourse and the Fly CLI can be found in [the official
documentation][concourse-docs].

### Zero-Downtime Deployment

The deployment tasks use the CloudFoundry v3 API commands in order to
support zero-downtime deployment and various new features.

When working with the cf cli tool, use the `v3` prefixed commands.
For example: `cf v3-apps`.

### Cancelling a failed deployment

Occasionally a zero-downtime deployment may fail.  This can occur if
the application fails to start, or the required number of app
instances cannot be created in time.  To cancel and roll back to the
previous deployment, run: `cf v3-cancel-zdt-push <app-name>`.

### Links to Staging

- [`govuk-coronavirus-business-volunteer-form`](https://govuk-coronavirus-business-volunteer-form-stg.cloudapps.digital/medical-equipment)
- [`govuk-coronavirus-vulnerable-people-form`](https://govuk-coronavirus-vulnerable-people-form-stg.cloudapps.digital/live-in-england)

These are protected by HTTP basic auth.  You can find the credentials
in govuk-secrets and get them by running:

```
cd ~/govuk/govuk-secrets/pass
./edit.sh 2ndline coronavirus-forms/creds
```

## Monitoring

> Grafana (there's a separate Grafana to look at) *TODO: where?*

Reviewer 6: I think this line could be removed.

Reviewer 3: We have Splunk. Splunk is good.

Reviewer 2: Cool, this is resolved

> 

Reviewer 3: ```suggestion
```

> 

Reviewer 3: ```suggestion
```

Reviewer 2: Looks like [this now a list](https://github.com/alphagov/govuk-developer-docs/pull/2423/files#diff-bc4c03cb61cd386ecda024855033dc90R243-R244)

There's a [Splunk dashboard][splunk] for both of these services.  To
access Splunk, you need to have the `GDS-006-GOVUK` permission on your
Google account.  To get this permission, raise an IT Helpdesk ticket
and post in `#cyber-security-help` to get them to confirm the request
is legitimate.

Application errors are sent to the GOV.UK hosted Sentry:

- [`govuk-coronavirus-business-volunteer-form`](https://sentry.io/organizations/govuk/issues/?project=5172100)
- [`govuk-coronavirus-vulnerable-people-form`](https://sentry.io/organizations/govuk/issues/?project=5170680)

These are under the GOV.UK Sentry account and you can sign in with
your Google Account.

Logs for both applications are streamed to Logit and can be found in
following dashboards:

GOV.UK Production Corona Forms *TODO: where?*

> GOV.UK Staging Corona Forms *TODO: where?*

Reviewer 6: We could provide a link to our Logit directly, from which folks can click through to these Kibana instances.

Reviewer 2: Looks like [links have been added](https://github.com/alphagov/govuk-developer-docs/pull/2423/files#diff-bc4c03cb61cd386ecda024855033dc90R252-R253)

## Extracting form responses (business only)

Cabinet Office will require data exports at regular intervals.

There is a rake task to export the data for a single day in JSON
format.  This can then be sent to the relevant team (TODO: who?).

```
cf v3-ssh govuk-coronavirus-business-volunteer-form
$ /tmp/lifecycle/shell
$ rake export:form_responses["<date>"]
```

Date to be included in the format 2020-03-26.

## Troubleshooting

### What things will call you

> *TODO: what?*

Reviewer 6: The GOV.UK PagerDuty on-call/2nd line pager will page you should these applications have issues.

Author: I assumed this section would be more "what sort of things will call you"

Reviewer 3: ```suggestion
The GOV.UK PagerDuty will page on-call/2ndline if these applications go down. It's connected up to a Pingdom check in the GOV.UK account that checks if the first form page is up.
```

applications inside.

#### Log in

Before you start, [install the CloudFoundry CLI][cf-cli].

1. Log in with `cf login -a api.cloud.service.gov.uk --sso`
2. Select staging or production with `cf target -s <staging or production>`

Both applications are Rails applications, and use the
[`ruby_buildpack`][] to build and deploy to PaaS.  You can read more
about [managing PaaS applications][paas-managing] and [monitoring PaaS
applications][paas-monitoring].

#### Manage backing services

The business volunteer application uses PaaS's [Redis backing
service][paas-redis].  The backing service is attached to the
application by a setting in the `manifest.yml` file in the repository
root.

The JSON data in the `VCAP_SERVICES` environment variable exposes
settings and credentials for the backing service automatically.

### AWS

We have two AWS accounts (staging and production) for CDN and data
storage. They contain DynamoDB, IAM and CloudFront resources,
which were provisioned by Terraform in the [tech-ops-private
repository](techops-repo).

To log in to AWS:

1. Install `gds-cli` v2.9.0 or above:

   ```
   brew update && brew upgrade gds-cli
   ```

> 2. Log in to the AWS console with either:

Reviewer 3: ```suggestion
1. [Install and set up the gds-cli](/manual/get-started.html#3-install-gds-tooling), then log in to the AWS Console with either:
```

For both applications, when the user submits the form on the "Check
your answers" page, the application writes user data to the database.

The vulnerable people form application only has permission to write
items to the database. For security and privacy reasons, there's no
way to read or change already-submitted user data.

Developers should not have access to the production database for

> the vunerable people form application.

Reviewer: ```suggestion
the vulnerable people form application.
```

Reviewer 2: Sorry sorry, really need to get my .md spellcheck back up and running... will rebase

> The business voulenteering form application can both read and write

Reviewer: ```suggestion
The business volunteering form application can both read and write
```

Reviewer 2: fixed! will get spellcheck for the code editor!

to it's database as the security and privacy requirements are lower.

> Developers may treat data in the business voulenteering form to the

Reviewer: ```suggestion
Developers may treat data in the business volunteering form to the
```

Reviewer 2: fixed! will get spellcheck for the code editor!

GOV.UK Production Corona Forms *TODO: where?*
GOV.UK Staging Corona Forms *TODO: where?*

## Extracting form responses (business only)

> Cabinet Office will require data exports at regular intervals.

Reviewer 3: Can we make this explicitly reference the "business volunteer form"? There are MUCH more stringent data controls around who has access to the vulnerable people data - I don't know what those are, but we shouldn't document them - if you know you know, you know?

Reviewer 2: I'll track down some answers for this one.

Reviewer 2: looks like that is amended in the title?
